src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs

```diff
@@ -1213,13 +1211,6 @@ public async Task GetAsync_AllowAutoRedirectTrue_NonRedirectStatusCode_LocationH
         public async Task GetAsync_AllowAutoRedirectTrue_RetainsOriginalFragmentIfAppropriate(
             string origFragment, string redirFragment, string expectedFragment, bool useRelativeRedirect)
         {
-            if (IsCurlHandler)
-            {
-                // Starting with libcurl 7.20, "fragment part of URLs are no longer sent to the server".
```

Comment:
@stephentoub i misread or this comment is not correct?Seems that curl behave like winhttp...this message states that fragment is not passed to server...but redirect test should work, right?

https://mc.dot.net/#/user/MarcoRossignoli/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/fa6463c63887d35a29f3e0485b5e18c3e413dff1/workItem/System.Net.Http.Functional.Tests/analysis/xunit/System.Net.Http.Functional.Tests.PlatformHandler_HttpClientHandlerTest~2FGetAsync_AllowAutoRedirectTrue_RetainsOriginalFragmentIfAppropriate(origFragment:%20%5C%22%23origFragment%5C%22,%20redirFragment:%20%5C%22%23redirFragment%5C%22,%20expectedFragment:%20%5C%22%23redirFragme)

Reply:
The assert that's failing isn't about what's sent to the server; it's about what the client-side HttpRequestMessage.RequestUri is getting set to:
https://github.com/dotnet/corefx/pull/32574/files#diff-06a6483ff6bb478e06f199d69522bcd4R1261

Reply:
You're right, i'll fix it, thank's!

Reply:
@stephentoub  i did some debug...and if i understand well the issue with CurlHandler is that there is no code that change `response.RequestMessage.RequestUri` before return in case o redirect.
On RedirectHandler you should set here https://github.com/dotnet/corefx/blob/3dbc455fb6f67be666541241e21c1218451af5a4/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RedirectHandler.cs#L64
Curl redirects automatically
Maybe we need to call https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs#L972 before https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs#L153 and add fragments if present.
This issue never rise up because this test was skipped for CurlHandler.
What do you think?

Reply:
>Curl redirects automatically

FYI.  WinHttpHandler also uses native WinHTTP to handle redirection automatically.  But we receive notifications from WinHTTP about this so we can update `request.RequestUri` as well as update cookies to the new Uri etc.

Reply:
Thank's @davidsh i'll take a look on it!

Reply:
@davidsh i found where we replace url on WinHttp https://github.com/dotnet/corefx/blob/master/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs#L208
and on Curl there is already a way to replace redirected url https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs#L982 calls https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs#L975 
From guide https://curl.haxx.se/libcurl/c/CURLINFO_EFFECTIVE_URL.html
```
In cases when you've asked libcurl to follow redirects, it may very well not be the same value you set with CURLOPT_URL.
```
However both doesn't handle fragment like SocketHttpHandler. If i remove WinHttp skip i get error(better than curl however that fails all cases):
```
System.Net.Http.Functional.Tests.PlatformHandler_HttpClientHandlerTest.GetAsync_AllowAutoRedirectTrue_RetainsOriginalFragmentIfAppropriate(origFragment: "#origFragment", redirFragment: "", expectedFragment: "#origFragment", useRela
  tiveRedirect: False) [FAIL]
        Assert.Equal() Failure
                                         ↓ (pos 23)
        Expected: ···p://127.0.0.1:11568/#origFragment
        Actual:   ···p://127.0.0.1:11568/
                                         ↑ (pos 23)
        Stack Trace:
          D:\git\corefx\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.cs(1261,0): at System.Net.Http.Functional.Tests.HttpClientHandlerTest.<>c__DisplayClass65_1.<<GetAsync_AllowAutoRedirectTrue_RetainsOriginalFragmentI
  fAppropriate>b__0>d.MoveNext()
          --- End of stack trace from previous location where exception was thrown ---
          D:\git\corefx\src\Common\tests\System\Net\Http\LoopbackServer.cs(66,0): at System.Net.Test.Common.LoopbackServer.CreateServerAsync(Func`2 funcAsync, Options options)
          D:\git\corefx\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.cs(1232,0): at System.Net.Http.Functional.Tests.HttpClientHandlerTest.GetAsync_AllowAutoRedirectTrue_RetainsOriginalFragmentIfAppropriate(String orig
  Fragment, String redirFragment, String expectedFragment, Boolean useRelativeRedirect)
          --- End of stack trace from previous location where exception was thrown ---
        Assert.Equal() Failure
      System.Net.Http.Functional.Tests.PlatformHandler_HttpClientHandlerTest.GetAsync_AllowAutoRedirectTrue_RetainsOriginalFragmentIfAppropriate(origFragment: "#origFragment", redirFragment: "", expectedFragment: "#origFragment", useRela
  tiveRedirect: True) [FAIL]
                                         ↓ (pos 23)
        Expected: ···p://127.0.0.1:11582/#origFragment
        Actual:   ···p://127.0.0.1:11582/
                                         ↑ (pos 23)
        Stack Trace:
          D:\git\corefx\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.cs(1261,0): at System.Net.Http.Functional.Tests.HttpClientHandlerTest.<>c__DisplayClass65_1.<<GetAsync_AllowAutoRedirectTrue_RetainsOriginalFragmentI
  fAppropriate>b__0>d.MoveNext()
          --- End of stack trace from previous location where exception was thrown ---
          D:\git\corefx\src\Common\tests\System\Net\Http\LoopbackServer.cs(66,0): at System.Net.Test.Common.LoopbackServer.CreateServerAsync(Func`2 funcAsync, Options options)
          D:\git\corefx\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.cs(1232,0): at System.Net.Http.Functional.Tests.HttpClientHandlerTest.GetAsync_AllowAutoRedirectTrue_RetainsOriginalFragmentIfAppropriate(String orig
  Fragment, String redirFragment, String expectedFragment, Boolean useRelativeRedirect)
          --- End of stack trace from previous location where exception was thrown ---
```
Maybe could be better skip failing assertion https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs#L1270 for Curl and WinHttp(as before) and discuss if align behaviour to SocketHttpHandler that respect all cases.

Reply:
I don't understand what you are proposing.

Before your PR, this test was always skipped on CurlHandler and WinHttpHandler.

Your PR changed that to run the test of CurlHandler.  But still skip it on WinHttpHandler.

Are you proposing that the test should go back to skipping on CurlHandler?  And it already skips on WinHttpHandler.

Reply:
> Are you proposing that the test should go back to skipping on CurlHandler? And it already skips on WinHttpHandler.

Yes, i did other tests and [WinHttp]( https://github.com/dotnet/corefx/blob/master/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs#L88) append fragment on redirect but [Curl](https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs#L975) not.

Reply:
We are not fixing CurlHandler.  So, the behavior is what it is.  So, you should put back the test code that skipped this test on CurlHandler.

The main point of this PR is to modify SocketsHttpHandler behavior to converge on the "correct" behavior long-term.

Reply:
ok...so i revert Curl skip...and update comment, now we skip due to behavior or redirect. Thank's for help!